src/logger/Logger.cpp

```diff
@@ -0,0 +1,194 @@
+#include "kompute/logger/Logger.hpp"
+
+#include <cassert>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <mutex>
+#include <spdlog/async.h>
+#include <spdlog/common.h>
+#include <spdlog/logger.h>
+#include <spdlog/sinks/rotating_file_sink.h>
+#include <spdlog/sinks/stdout_color_sinks.h>
+#include <spdlog/spdlog.h>
+
+#include <sys/stat.h>
+#include <sys/types.h>
+
+#ifdef _WIN32
+#include <iso646.h>
+#include <direct.h>
+#endif // _WIN32
+
+namespace logger {
+constexpr int THREAD_QUEUE_LENGTH = 8192;
+constexpr int FILE_ROTATION_TIME = 1048576 * 5;
+
+/**
+ * Should be replaced with:
+ * std::filesystem::exists(path)
+ * when switching to cpp17
+ **/
+bool
+exists(const std::string& path)
+{
+    struct stat info
+    {};
+    if (stat(path.c_str(), &info) != 0) {
+        // std::cerr << "Failed to check if '" << path
+        //           << "' exists. Cannot access!\n";
+        // assert(false);
+        return false;
+    }
+    return info.st_mode & S_IFDIR;
+}
+
+/**
+ * Based on: https://stackoverflow.com/a/35109823
+ * Should be replaced with:
+ * std::filesystem::create_directory(path);
+ * when switching to cpp17
+ **/
+void
+createDir(const std::string& path)
+{
+    int nError = 0;
+#if defined(_WIN32)
+    nError = _mkdir(path.c_str()); // can be used on Windows
+#else
+    mode_t nMode = 0733; // UNIX style permissions
+    nError = mkdir(path.c_str(), nMode); // can be used on non-Windows
+#endif
+    if (nError != 0) {
+        std::cerr << "Failed to create '" << path << "' with: " << nError
+                  << '\n';
+        assert(false);
+    }
+}
+
+void
+setupLogger()
+{
+    // Ensure we setup the logger only once
+    static bool setup = false;
+    static std::mutex setupMutex{};
+    setupMutex.lock();
+    if (setup) {
+        setupMutex.unlock();
+        return;
+    }
+    setup = true;
+    setupMutex.unlock();
+
+    if (!exists(logger::logFolder)) {
+        createDir(logger::logFolder);
+    }
+    spdlog::init_thread_pool(THREAD_QUEUE_LENGTH, 1);
+    spdlog::sink_ptr console_sink =
+      std::make_shared<spdlog::sinks::stdout_color_sink_mt>();
+    console_sink->set_pattern("[%H:%M:%S %z] [%=8l] [thread %t] [%@]\t%v");
+#ifdef _WIN32
+    std::string s = logger::logFolder + "\\kompute.log";
+    spdlog::sink_ptr file_sink =
+      std::make_shared<spdlog::sinks::rotating_file_sink_mt>(
+        s, FILE_ROTATION_TIME, 3);
+#else // _WIN32
+    spdlog::sink_ptr file_sink =
+      std::make_shared<spdlog::sinks::rotating_file_sink_mt>(
+        logger::logFolder + "/kompute.log", FILE_ROTATION_TIME, 3);
+#endif
+    file_sink->set_pattern("[%H:%M:%S %z] [%=8l] [thread %t] [%@]\t%v");
+    std::vector<spdlog::sink_ptr> sinks{ file_sink, console_sink };
+    std::shared_ptr<spdlog::logger> logger =
```

Comment:
What is the reason for us to implement our own threaded based logging functionality in the main package, seems quite opinionated and often people won't want to use spdlog in their projects or may want to use it in a different way

Reply:
I don't think we should add these helpers on the core library as this should ideally be in the application code if people do want it and can follow spdlog specific implementation if they want to use spdlog https://github.com/gabime/spdlog#rotating-files - i added comments above so resolving this one so we can discuss in that other one